BungeeCord-Patches/0056-Save-memory-on-scoreboard-teams.patch

```diff
@@ -0,0 +1,107 @@
+From 79a63969805be4e5d444eb4f7360141de14176d7 Mon Sep 17 00:00:00 2001
+From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
+Date: Fri, 8 May 2020 21:39:39 +0300
+Subject: [PATCH] Save memory on scoreboard teams
+
+
+diff --git a/api/src/main/java/net/md_5/bungee/api/score/Team.java b/api/src/main/java/net/md_5/bungee/api/score/Team.java
+index 39f81fd1..093784f1 100644
+--- a/api/src/main/java/net/md_5/bungee/api/score/Team.java
++++ b/api/src/main/java/net/md_5/bungee/api/score/Team.java
+@@ -17,8 +17,8 @@ public class Team
+     private String prefix;
+     private String suffix;
+     private byte friendlyFire;
+-    private String nameTagVisibility;
+-    private String collisionRule;
++    private NametagVisibility nameTagVisibility; // Waterfall - save memory on scoreboard teams
++    private CollisionRule collisionRule; // Waterfall - save memory on scoreboard teams
+     private int color;
+     private Set<String> players = new HashSet<>();
+ 
+@@ -32,6 +32,82 @@ public class Team
+         players.add(name.intern());
+     }
+ 
++    // Waterfall start - save memory on scoreboard teams
++    @Deprecated
++    public String getNameTagVisibility() {
++        return nameTagVisibility.getValue();
++    }
++
++    @Deprecated
++    public void setNameTagVisibility(String value) {
++        NametagVisibility enumValue;
++        if (value == null) {
++            enumValue = NametagVisibility.ALWAYS;
++        } else if ((enumValue = NametagVisibility.VALUES.get(value)) == null) {
++            net.md_5.bungee.api.ProxyServer.getInstance().getLogger().log(java.util.logging.Level.WARNING, "Unknown team name tag visibility value: {0} -> {1}", new Object[] { name, value });
++            enumValue = NametagVisibility.ALWAYS;
++        }
++        this.nameTagVisibility = enumValue;
++    }
++
++    @Deprecated
++    public String getCollisionRule() {
++        return collisionRule.getValue();
++    }
++
++    @Deprecated
++    public void setCollisionRule(String value) {
++        CollisionRule enumValue;
++        if (value == null) {
++            enumValue = CollisionRule.ALWAYS;
++        } else if ((enumValue = CollisionRule.VALUES.get(value)) == null) {
++            net.md_5.bungee.api.ProxyServer.getInstance().getLogger().log(java.util.logging.Level.WARNING, "Unknown team collision rule value: {0} -> {1}", new Object[] { name, value });
++            enumValue = CollisionRule.ALWAYS;
++        }
++        this.collisionRule = enumValue;
++    }
++
++    @lombok.RequiredArgsConstructor
++    public enum CollisionRule {
++        ALWAYS("always"),
++        PUSH_OTHER_TEAMS("pushOtherTeams"),
++        PUSH_OWN_TEAM("pushOwnTeam"),
++        NEVER("never"),
++        ;
++
++        @lombok.Getter
++        private final String value;
++
++        public static final java.util.Map<String, CollisionRule> VALUES = new java.util.HashMap<>();
```

Comment:
Suggesting to use ImmutableMap or sth. similar

Reply:
Eh i don't think the complexity in the change is worth that. immutable isnt gonna offer any real performance benefit.

Reply:
Yeah, if there are 10 fields more or less in that array it won't make any difference in performance, I just thought it'd look better.